### DISCUSSION TITLE
Stackdriver Logging datasource

### DISCUSSION THREAD

**Discussion by rdigiorgio** [grafana/grafana#41820] [Data Sources]
**What would you like to be added**:

As of now we are able to query and display Stackdriver Metrics into Grafana, I think it would be great to also be able to query and display Stackdriver Logs. I am actually surprised that I was not able to find any issue or pull request about this already.

**Why is this needed**:

This is needed because I want to have my metrics and logs in the same place, making it possible to make dashboards where I could display metrics and logs for a given Stackdriver resource.

**Comment by aocenas** [comment]
Hey @sunker I think you worked on stackdriver a lot recently, do you know if there are plans to do this?

**Comment by sunker** [comment]
No there's no plans to do this at the moment, but I think it's great idea.

**Comment by sunker** [comment]
Questions: 

- Would we want to support both basic and advanced search mode in the query editor? It seems like the basic mode is an abstraction on top of advanced mode (it's using the same api as the advanced mode query), so therefore it would be easier for us to just implement advanced mode at first.
- There seems to be a new version of the Logs Viewer coming soon. Is that UI using the same underlying api as the classic Logs Viewer? For some reason, the option to preview the new logs viewer is not available to me. 

![Skärmavbild 2020-02-28 kl  11 48 11](https://user-images.githubusercontent.com/2388950/75542792-c774df00-5a20-11ea-90de-7d8766463899.png)

- In the Logs Viewer it's possible to stream logs. Is there a public api for that? How does it work - polling or web sockets? 
- For Loki there's a feature called [`Show Context `](https://grafana.com/docs/grafana/latest/features/datasources/loki/#log-context), which allows the user to query the surrounding lines of log entry. Would it be possible to do something similar for Stackdriver logs?
- How to link to Stackdriver traces from a log row (can we use the DerivedField feature)?
- Intellisense/suggestions: It says in the docs: `The suggestions come from both the query language and the set of actual log entries that the Logs Viewer has loaded`. Would it be possible to get access to a query language schema?
- At the moment Explore has support for expanding key-value pairs ( ... key=value ...) in the log message so that it can be displayed as parsed fields in the detailed view. Can it also parse json data? That will be necessary when handling structured logs for Stackdriver. Also wondering how Explore handles log messages that are nested json structures. 
- It seems to be possible to list all [logs within a project](https://cloud.google.com/logging/docs/reference/v2/rest/v2/projects.logs/list), but I cannot find an endpoint that lists all resource keys within a project. That will be necessary in order to provide intellisense/an interactive query editor with dropdowns.

**Comment by stefanforsberg** [comment]
Thank a lot for looking into this, would be great for our dashboards =)

I can't help with any of the interesting question but I'll chip 2c in anyway.

We are only using the advance mode and by your description it seems to make total sense to implement that (first).

When it comes to the new version of Log Viewer seems to be rolled out atm but be available for everyone by March 2nd.

**Comment by sunker** [comment]
Nice, thanks Stefan! Hopefully we can start development within the next couple of weeks. /Erik

**Comment by eli-jordan** [comment]
@sunker We are very interested in this capability too. We are currently planning to export logs to BigQuery then use the BQ plugin. If we could go directly to SD logging that would be great. Is there somewhere I can follow the progress?

**Comment by mtanda** [comment]
Cloud Logging will add new feature, it seems to be very useful.
I hope Grafana support these feature too.

`Log Analytics` provides analytics feature by using BigQuery.
https://cloud.google.com/logging/docs/log-analytics

`Live tailing` provides `tail for log` feature.
https://cloud.google.com/logging/docs/view/streaming-live-tailing

Google already provides API for it.
https://cloud.google.com/logging/docs/reference/v2/rest/v2/entries/tail

**Reply by mtanda** [reply]
`Log Analytics` provides BigQuery interface.
Grafana already provides [BigQuery datasource](https://grafana.com/grafana/plugins/grafana-bigquery-datasource/).
Now, we can display logs in Grafana!

**Reply by Passe12345** [reply]
So the only 2 options to view GCP logs in Grafana would be either to use BigQuery or having a Promtail/Grafana Agent already running for pushing logs to Loki. Right ?
Log Analytics restrictions and BQ doesn't seem appropriate for the purpose of viewing and manipulating logs in Grafana, unless i'm mistaking.

As a Grafana Cloud user it's frustating to have to deploy a dedicated Promtail instance for this purpose.

**Comment by nicolasgere** [comment]
Any update ?

**Comment by vickyyyyyyy** [comment]
Thanks for everyone's patience, we understand this has been a long requested feature and are currently working with GCP on a better solution. We don't have a timeframe for this yet but will share any updates as soon as we can!

**Comment by Traxmaxx** [comment]
I just found out the GCP team released something:
https://grafana.com/grafana/plugins/googlecloud-logging-datasource/
